transcendence-app/src/authorization/infrastructure/db/postgres/user-to-role.postgres.repository.ts

```diff
@@ -0,0 +1,102 @@
+import { table } from '../../../../shared/db/models';
+import { makeQuery } from '../../../../shared/db/postgres/utils';
+import { userKeys } from '../../../../user/infrastructure/db/user.entity';
+import { Role } from '../../../../shared/enums/role.enum';
+import {
+  UserWithAuthorization,
+  UserWithAuthorizationData,
+} from '../user-with-authorization.entity';
+
+@Injectable()
+export class UserToRolePostgresRepository
+  extends BasePostgresRepository<UserToRole>
+  implements IUserToRoleRepository
+{
+  constructor(protected pool: PostgresPool) {
+    super(pool, table.USER_TO_ROLE, UserToRole);
+  }
+
+  async addUserToRole(
+    userKey: string,
+    role: Role,
+    isUserId: boolean,
+  ): Promise<UserToRole | null> {
+    const userKeyMatcher = isUserId
+      ? `u.${userKeys.ID}`
+      : `u.${userKeys.USERNAME}`;
+    const userData = await makeQuery<UserToRole>(this.pool, {
+      text: `WITH myUser as (SELECT u.${userKeys.ID}
+                             FROM ${table.USERS} u
+                             WHERE ${userKeyMatcher} = $1)
+             INSERT
+             INTO ${this.table} (${UserToRoleKeys.ID}, ${UserToRoleKeys.ROLE})
+             VALUES ((SELECT ${userKeys.ID} from myUser), $2)
+             RETURNING *;`,
+      values: [userKey, role],
+    });
+    return userData && userData.length > 0 ? new UserToRole(userData[0]) : null;
+  }
+
+  async getUserWithAuthorization(
+    userKey: string,
+    isUserId: boolean,
+  ): Promise<UserWithAuthorization | null> {
+    const userKeyMatcher = isUserId
+      ? `u.${userKeys.ID}`
+      : `u.${userKeys.USERNAME}`;
+    const members = await makeQuery<UserWithAuthorizationData>(this.pool, {
+      text: `WITH UserWithRoles AS (SELECT u.${userKeys.ID},
+                                           u.${userKeys.USERNAME},
+                                           coalesce(array_agg(ur.${UserToRoleKeys.ROLE})
+                                                    FILTER
+                                                      (WHERE ur.${UserToRoleKeys.ROLE} IS NOT NULL),
+                                                    '{}') as roles
+                                    FROM ${table.USERS} u
+                                           LEFT JOIN ${this.table} ur ON ur.${UserToRoleKeys.ID} = u.${userKeys.ID}
+                                    WHERE ${userKeyMatcher} = $1
+                                    GROUP BY u.${userKeys.ID})
+             SELECT ${userKeys.ID}                                                          as userId,
+                    ${userKeys.USERNAME},
+                    to_jsonb((SELECT roles from UserWithRoles)) @> '["owner"]'::jsonb       as g_owner,
+                    to_jsonb(((SELECT roles from UserWithRoles))) @> '["moderator"]'::jsonb as g_admin,
+                    to_jsonb(((SELECT roles from UserWithRoles))) @> '["banned"]'::jsonb    as g_banned
```

Comment:
question: What is the reason we convert the array to `jsonb`?
The `@>` operator seems to work with arrays:

```sql
                    (SELECT roles from UserWithRoles) @> '{"owner"}'      as g_owner,
                    (SELECT roles from UserWithRoles) @> '{"moderator"}'  as g_admin,
                    (SELECT roles from UserWithRoles) @> '{"banned"}'     as g_banned
```

Reply:
You are right. I got something wrong at some point.
Fixed.